### PR TITLE
Make deadlocks less cryptic, cut rate of aborted transactions

### PR DESCRIPTION
**Pull request by jakewins**
Two changes; one, each enterprise lock client now has a description of what it is doing (eg. a cypher query, generally), which it outputs in debugging and error outputs. This will help correlate locking problems with the queries involved in them.

Two, change the enterprise lock resolution strategy to abort the transaction with the lowest number of held locks, rather than the transaction with the shortest wait list. The old strategy was not strictly flawed, but severely inefficient. The included java program, which tests concurrent transactions acquiring four locks each, shows the number of aborted transactions dropped by 40%(!) after this change.

### CODE REVIEW

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/DefaultDeadlockStrategy.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package org.neo4j.kernel.ha.lock.forseti;
+
+/**
+ * The standard deadlock strategy. When a deadlock occurs, the client with the fewest number of held locks is aborted. If both clients hold the same number of
+ * locks, the client with the lowest client id is aborted.
+ *
+ * This is one side of a long academic argument, where the other says to abort the one with the most locks held, since it's old and monolithic and holding up
+ * the line.
+ */
+public class DefaultDeadlockStrategy implements ForsetiLockManager.DeadlockResolutionStrategy
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/DefaultDeadlockStrategy.java]: This class is missing a license header.

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/lock/forseti/ForsetiDeadlockResolutionExploration.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import org.neo4j.kernel.DeadlockDetectedException;
+import org.neo4j.kernel.impl.locking.Locks;
+import org.neo4j.kernel.impl.locking.ResourceTypes;
+
+import static java.util.Arrays.asList;
+
+/**
+ * Kept for future reference. A simple program that concurrently acquires and releases locks and tracks the number of deadlocks that occur.
+ */
+public class ForsetiDeadlockResolutionExploration
```
Comment (chrisvest) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/lock/forseti/ForsetiDeadlockResolutionExploration.java]: This class is also missing a license header.

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/lock/forseti/ForsetiMessagesTest.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import org.neo4j.kernel.DeadlockDetectedException;
+import org.neo4j.kernel.impl.locking.Locks;
+import org.neo4j.kernel.impl.locking.ResourceTypes;
+import org.neo4j.test.OtherThreadExecutor;
+import org.neo4j.test.OtherThreadRule;
+
+import static org.junit.Assert.assertEquals;
+
+/** Test output of various Forseti introspection mechanisms */
+public class ForsetiMessagesTest
```
Comment (chrisvest) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/lock/forseti/ForsetiMessagesTest.java]: This one also doesn't have a license header.

**Comment by davidegrohmann** [comment]
@jakewins this looks a quite big change and it is also pretty much untested. How do we know that this is an improvement?

Can we perhaps transform the manual test program in a well written stress test where to assert something smart?

**Comment by jakewins** [comment]
@davidegrohmann on the commit to include a description I agree there should be more tests, I'll pop some in. 

On the main commit, deadlock resolution resolution, it's not changing surface contracts, just improving our fulfilment of them. Ergo, the existing locking test suites already cover this behavior and these code paths, so I don't agree there is no test coverage. 

I see your point about making the test tool I wrote into a real test. We'd need to use probabilistic assertions for that (at most x% of attempts should deadlock), which is not ideal.. but it would be nice to lock (heh) this behavior in place so we don't regress. I'm wondering if we could simply add an assertion to the locking stress test.. I'll have a look

**Comment by jakewins** [comment]
@davidegrohmann @chrisvest updated PR by adding deadlock frequency assertion to the locking fuzz test, and freshening that up while I was down there.

Some interesting things found: The change has no noticeable effect on that test, the deadlock frequency remains the same. I'm still able to reproduce the exact same benefit in the original "manual" test, the two tests are very similar, and I've yet to sort out what causes the difference.

Second; there's a discrepancy in the deadlock frequency in Forseti vs. the old lock manager in that test - the old lock manager sees significantly fewer deadlocks. I've spent the afternoon trying to sort out why - worrying that it's to do with the dreadlocks detection algorithm and its false-positive chance. However, adding manual checks to double-check in synchronized blocks that it's detecting real deadlocks made that seem unlikely.. One notable difference is that Forseti is faster (2s vs 44s), so maybe it's simply stress threads "catching up" to each other, making real deadlocks more common. In any case, something worth investigating further as time permits.

**Comment by jakewins** [comment]
Lets see how that last update works. In other news, we really should abandon the dreadlocks algo, it is such a mess to debug. We should be able to get a lock manager to scale without having to deal with the complexity of eventually consistent wait lists..

**Comment by jakewins** [comment]
@chrisvest updated again, this is starting to get embarrasing.. :p What are you doing reviewing this now anyway when you should be busy joining https://oreslack.herokuapp.com/

**Comment by jakewins** [comment]
@chrisvest to the merge mobile?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java [hunk 4]
```diff
@@ -42,6 +42,18 @@ public CommunityLockClient( LockManagerImpl manager )
     }
 
     @Override
+    public Locks.Client description( String desc )
+    {
+        return this;
+    }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java]: Don't the community lock manager need better descriptions?

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/ForsetiClient.java [hunk 5]
```diff
@@ -73,20 +88,28 @@
 
     /**
      * For exclusive locks, we only need a single re-usable one per client. We simply CAS this lock into whatever slots
-     * we want to hold in the global lock map. */
+     * we want to hold in the global lock map.
+     */
     private final ExclusiveLock myExclusiveLock = new ExclusiveLock(this);
 
+    private final CyclicBarrier deadlockVerificationBarrier = new CyclicBarrier( 2 );
+
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/ForsetiClient.java]: This guy's never used.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/SharedLock.java [hunk 6]
```diff
@@ -117,26 +118,7 @@ public void copyHolderWaitListsInto( SimpleBitSet waitList )
...
-                    size += client.waitListSize();
-                }
-            }
-        }
-        return size;
-    }
-
-    @Override
-    public boolean anyHolderIsWaitingFor( int clientId )
+    public int detectDeadlock( int clientId )
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/forseti/SharedLock.java:121]: Nice.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/ThreadToStatementContextBridge.java [hunk 7]
```diff
@@ -60,10 +60,16 @@ public void unbindTransactionFromCurrentThread()
...
     public Statement instance()
     {
+        return tx().acquireStatement();
+    }
+
+    public KernelTransaction tx()
+    {
         checkIfShutdown();
-        return getKernelTransactionBoundToThisThread( true ).acquireStatement();
+        return getKernelTransactionBoundToThisThread( true );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/ThreadToStatementContextBridge.java]: I don't like the `tx` method name. This method is part of the `get*TransactionBoundToThisThread` family of methods.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/collection/TimedRepository.java [hunk 8]
```diff
@@ -44,11 +42,16 @@
...
-    private final Factory<VALUE> factory;
+    private final ValueFactory<KEY, VALUE> factory;
     private final Consumer<VALUE> reaper;
     private final long timeout;
     private final Clock clock;
 
+    public interface ValueFactory<KEY, VALUE>
+    {
+        VALUE newInstance( KEY key );
+    }
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/collection/TimedRepository.java]: This is the same as the Function interface.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/MasterClient210.java [hunk 9]
```diff
@@ -225,7 +225,7 @@ public Integer read( ChannelBuffer buffer, ByteBuffer temporaryBuffer ) throws I
     public Response<LockResult> acquireExclusiveLock( RequestContext context, Locks.ResourceType type, long...
             resourceIds )
     {
-        return sendRequest( HaRequestType210.ACQUIRE_EXCLUSIVE_LOCK, context,
+            return sendRequest( HaRequestType210.ACQUIRE_EXCLUSIVE_LOCK, context,
                 new AcquireLockSerializer( type, resourceIds ), LOCK_RESULT_DESERIALIZER );
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/MasterClient210.java]: Some accidental indentation.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/com/master/MasterImpl.java [hunk 10]
```diff
@@ -144,11 +143,12 @@ private long generateEpoch()
...
     {
-        this.slaveLockSessions = new TimedRepository<>( new Factory<Locks.Client>()
+        this.slaveLockSessions = new TimedRepository<>( new TimedRepository.ValueFactory<RequestContext, Locks.Client>()
             {
-                @Override public Locks.Client newInstance()
+                @Override public Locks.Client newInstance(RequestContext ctx)
                 {
-                    return spi.acquireClient();
+                    return spi.acquireClient().description(
+                            String.format("Locks held on behalf of slave `%d`, slave transaction id `%d`", ctx.machineId(), ctx.getEventIdentifier() ) );
```
Comment (chrisvest) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/com/master/MasterImpl.java]: This needs formatting. The line is too long.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java (hunk 4 above)
Comment (jakewins) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/locking/community/CommunityLockClient.java]: The community lock manager prints the resource access graph on deadlock, so it's much more helpful already. Agree that these descriptions would be really nice in community lock manager as well though, vote we do that in a follow up commit at some point

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/ThreadToStatementContextBridge.java (hunk 7 above)
Comment (jakewins) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/ThreadToStatementContextBridge.java]: Doh, I didn't realize there was a public accessor already.. removing this method altogether

**Comment by tinwelint** [comment]
@jakewins @chrisvest what's the state of this PR?

**Comment by jakewins** [comment]
It should be ready to rock, assuming @chrisvest is happy with the updates :)

**Pull request merged**